src/mlpack/tests/main_tests/sparse_coding_test.cpp

```diff
@@ -0,0 +1,746 @@
+/**
+ * @file sparse_coding_test.cpp
+ * @author Manish Kumar
+ *
+ * Test mlpackMain() of sparse_coding_main.cpp.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <string>
+
+#define BINDING_TYPE BINDING_TYPE_TEST
+static const std::string testName = "SparseCoding";
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/methods/sparse_coding/sparse_coding_main.cpp>
+#include "test_helper.hpp"
+
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+using namespace mlpack;
+
+struct SparseCodingTestFixture
+{
+ public:
+  SparseCodingTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~SparseCodingTestFixture()
+  {
+    // Clear the settings.
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(SparseCodingMainTest, SparseCodingTestFixture);
+
+/**
+ * Make sure that output points in dictionary equals number of
+ * atoms passed and codes have desired dimension.
+ */
+BOOST_AUTO_TEST_CASE(SparseCodingOutputDimensionTest)
+{
+  arma::mat inputData;
+  if (!data::Load("iris_train.csv", inputData))
```

Comment:
Just an idea, maybe it makes sense to write an extra function for the dataset loading since it's basically used in every test. What do you think?

Reply:
Thanks @zoq for reviewing and suggestion. I like your idea to make a dataset loader function and tried to implement it.
I tried to design it in a way so that it can load training data, training labels, test data, test labels and training weights depending on data requirement.
Below is the working code tested over sparse coding bindings tests:
```
template<typename T  = arma::mat , typename G = T, typename W = G>
void LoadData(T& trainData = *(T*) NULL,
              const std::string& trainPath = *(std::string*) NULL,
              T& testData = *(T*) NULL,
              const std::string& testPath = *(std::string*) NULL,
              G& trainLabels = *(G*) NULL,
              const std::string& trainLabelsPath = *(std::string*) NULL,
              G& testLabels = *(G*) NULL,
              const std::string& testLabelsPath = *(std::string*) NULL,
              W& weights = *(W*) NULL,
              const std::string& weightsPath = *(std::string*) NULL)
{
  // Load training data.
  if(&trainPath != NULL)
    if(!data::Load(trainPath, trainData))
      Log::Fatal << "Cannot load train dataset" << trainPath << "!"
          << std::endl;

  // Load test data.
  if(&testPath != NULL)
    if(!data::Load(testPath, testData))
      Log::Fatal << "Cannot load test dataset" << testPath << "!"
          << std::endl;

  // Load training data labels.
  if(&trainLabelsPath != NULL)
    if(!data::Load(trainLabelsPath, trainLabels))
      Log::Fatal << "Cannot load train labels" << trainLabelsPath << "!"
          << std::endl;

    // Load test data labels.
  if(&testLabelsPath != NULL)
    if(!data::Load(testLabelsPath, testLabels))
      Log::Fatal << "Cannot load test labels" << testLabelsPath << "!"
          << std::endl;

    // Load test data labels.
  if(&weightsPath != NULL)
    if(!data::Load(weightsPath, weights))
      Log::Fatal << "Cannot load training weights" << weightsPath << "!"
          << std::endl;
}
```
Implementation may seem a bit weird as I couldn't find anything else to assign default values to the arguments passed by reference.
Please provide your views on it.

Reply:
I think this could be useful, but I'd consider not using default arguments at all, but instead offering a few overloads:

```
template<typename MatType, typename LabelsType, typename WeightsType>
void LoadData(const std::string& trainFile,
              const std::string& trainLabelsFile,
              const std::string& testFile,
              const std::string& testLabelsFile,
              const std::string& trainWeightsFile,
              MatType& trainData,
              LabelsType& trainLabels,
              MatType& testData,
              LabelsType& testLabels,
              WeightsType& trainWeights);

template<typename MatType, typename LabelsType>
void LoadData(const std::string& trainFile,
              const std::string& trainLabelsFile,
              const std::string& testFile,
              const std::string& testLabelsFile,
              MatType& trainData,
              LabelsType& trainLabels,
              MatType& testData,
              LabelsType& testLabels);

template<typename MatType, typename LabelsType>
void LoadData(const std::string& trainFile,
              const std::string& trainLabelsFile,
              MatType& trainData,
              LabelsType& trainLabels);
```

(I ordered the parameters differently---the idea was all input parameters first, then all output parameters.)  That's just one possible idea.  I guess that it would be possible to generalize a little bit to pass in an arbitrary number of strings followed by an arbitrary number of objects to be loaded using variadic templates; that might result in the smallest implementation.

Reply:
I was thinking more in the context of this particular test:

```c++
void LoadData(arma::mat& inputData, arma::mat& testData)
{
  // Load train dataset.
  if (!data::Load("iris_train.csv", inputData))
    BOOST_FAIL("Cannot load train dataset iris_train.csv!");

  // Load test dataset.
  if (!data::Load("iris_test.csv", testData))
    BOOST_FAIL("Cannot load test dataset iris_test.csv!");
}
```

Reply:
Oh, fair, that works too. :)

Reply:
@rcurtin, @zoq : So, shall we generalize the data loading through overloading in test_helper.cpp as @rcurtin said or should we only employ the data loading for this particular test file as suggested by @zoq .

Reply:
Should be fine, to implement a simple ``LoadData`` function inside the test, we can implement a generalized function later, if we think that might be helpful. Does this sound good?